src/test/java/StringManipulationTest.java

```diff
@@ -118,32 +200,51 @@ public void testRestoreString1()
         assertEquals(restoreString, "rat");
     }
 
+    /**
+     * Check whether a simple string can be correctly rearranged into all the same letter (this wasn't part of the spec
+     * but I am making the executive decision to include it).
+     */
     @Test
     public void testRestoreString2()
     {
-        fail("Not yet implemented");
-
+        manipulatedstring.setString("abc");
+        int [] array = new int[]{2,2,2};
+        assertEquals(manipulatedstring.restoreString(array), "ccc");
     }
 
+    /**
+     * Check whether an illegal argument exception is correctly thrown when the number of indices doesn't match
+     * the number of letters in the string.
+     */
     @Test
     public void testRestoreString3()
     {
-        fail("Not yet implemented");
-
+        manipulatedstring.setString("abcd");
+        int [] array = new int[]{1,2,3};
+        assertThrows(IllegalArgumentException.class, () -> manipulatedstring.restoreString(array));
     }
-
+    /**
+     * Check whether an illegal argument exception is correctly thrown when one of the indices is higher than the
+     * string length.
+     */
     @Test
     public void testRestoreString4()
     {
-        fail("Not yet implemented");
-
+        manipulatedstring.setString("abc");
+        int [] array = new int[]{1,2,3};
+        assertThrows(IndexOutOfBoundsException.class, () -> manipulatedstring.restoreString(array));
     }
 
+    /**
+     * Check whether a string with non-alphanumeric characters can be successfully rearranged. (Also not part
+     * of the spec, but my implementation does it anyway and I wanted to test it.)
+     */
     @Test
     public void testRestoreString5()
```

Comment:
And this one, you're throwing in non-alphanumeric characters into the mix. Real-world strings can be messy and full of all sorts of characters, so it's awesome that you're making sure your method can handle that